File: study1/solutions/week3/keys_and_rooms.py
# https://leetcode.com/problems/keys-and-rooms/description/
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = set([0])
        to_visit = [0]

        while to_visit:
            current_node = to_visit.pop()
            for key in rooms[current_node]:
                if key not in visited:
                    to_visit.append(key)
                    visited.add(key)
        return len(visited) == len(rooms)

File: study1/week3/keys-and-rooms.java
// https://leetcode.com/problems/keys-and-rooms/
// Time Complexity: O(n^2)
// Space Complexity: O(n), n: rooms.size()
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        List<Integer> keyList = new ArrayList<>(); // key list which I have.
        boolean[] visited = new boolean[rooms.size()]; // if it is open, then true
        int roomIdx = 0;
        keyList.add(0);

        while (true) {
            System.out.println(roomIdx);
            if (!visited[roomIdx]) {
                // visit the room and pick up key
                keyList.addAll(rooms.get(roomIdx));
                // check unlocked room
                visited[roomIdx] = true;
            }
            keyList.remove(Integer.valueOf(roomIdx));
            if (!keyList.isEmpty()) {    // set the roomIdx to the next room which will be visited
                roomIdx = keyList.get(0);
            } else {
                break;
            }
        }

        for (boolean v : visited) {
            if (!v) {    // if there is room which we can not enter
                return false;
            }
        }

        return true;
    }
}